runtime/lib/backend/openqasm/OpenQasmDevice.cpp

```diff
@@ -0,0 +1,213 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "OpenQasmDevice.hpp"
+
+namespace Catalyst::Runtime::Device {
+
+auto OpenQasmDevice::AllocateQubit() -> QubitIdType
+{
+    RT_FAIL("Unsupported functionality");
+    return 0;
+}
+
+auto OpenQasmDevice::AllocateQubits(size_t num_qubits) -> std::vector<QubitIdType>
+{
+    if (num_qubits == 0U) {
+        return {};
+    }
+
+    const size_t cur_num_qubits = this->device->getNumQubits();
+    const size_t new_num_qubits = cur_num_qubits + num_qubits;
+    if (cur_num_qubits) {
+        this->device = std::make_unique<OpenQasm::OpenQasmBuilder>();
+    }
+
+    this->device->Register(OpenQasm::RegisterType::Qubit, "qubits", new_num_qubits);
+
+    return this->qubit_manager.AllocateRange(cur_num_qubits, new_num_qubits);
+}
+
+void OpenQasmDevice::ReleaseAllQubits()
+{
+    // do nothing
+}
+
+void OpenQasmDevice::ReleaseQubit([[maybe_unused]] QubitIdType q)
+{
+    RT_FAIL("Unsupported functionality");
+}
+
+auto OpenQasmDevice::GetNumQubits() const -> size_t { return this->device->getNumQubits(); }
+
+void OpenQasmDevice::StartTapeRecording() { RT_FAIL("Unsupported functionality"); }
+
+void OpenQasmDevice::StopTapeRecording() { RT_FAIL("Unsupported functionality"); }
+
+void OpenQasmDevice::SetDeviceShots([[maybe_unused]] size_t shots)
+{
+    RT_FAIL("Unsupported functionality");
+}
+
+auto OpenQasmDevice::GetDeviceShots() const -> size_t
+{
+    RT_FAIL("Unsupported functionality");
+    return 0;
+}
+
+void OpenQasmDevice::PrintState() { RT_FAIL("Unsupported functionality"); }
+
+auto OpenQasmDevice::Zero() const -> Result
+{
+    RT_FAIL("Unsupported functionality");
+    return 0;
+}
+
+auto OpenQasmDevice::One() const -> Result
+{
+    RT_FAIL("Unsupported functionality");
+    return 0;
+}
+
+void OpenQasmDevice::PrintCircuit() { std::cout << this->device->toOpenQasm(); }
+
+auto OpenQasmDevice::Circuit() -> std::string { return this->device->toOpenQasm(); }
+
+void OpenQasmDevice::NamedOperation(const std::string &name, const std::vector<double> &params,
+                                    const std::vector<QubitIdType> &wires, bool inverse)
+{
+    using namespace Catalyst::Runtime::Simulator::Lightning;
+
+    // First, check operation specifications
+    auto &&[op_num_wires, op_num_params] = lookup_gates(simulator_gate_info, name);
+
+    // Check the validity of number of qubits and parameters
+    RT_FAIL_IF((!wires.size() && wires.size() != op_num_wires), "Invalid number of qubits");
+    RT_FAIL_IF(params.size() != op_num_params, "Invalid number of parameters");
+
+    // Convert wires to device wires
+    auto &&dev_wires = getDeviceWires(wires);
+
+    this->device->Gate(name, params, {}, dev_wires, inverse);
```

Comment:
Hi Ali, could you please point me where to look for the implementation of this call?